src/t_zset.c

```diff
@@ -1965,17 +1965,20 @@ void zuiInitIterator(zsetopsrc *op) {
             serverPanic("Unknown set encoding");
         }
     } else if (op->type == OBJ_ZSET) {
+        /* Sorted sets are traversed in reverse order to optimize for
+         * the insertion of the elements in a new list as in
+         * ZDIFF/ZINTER/ZUNION */
         iterzset *it = &op->iter.zset;
         if (op->encoding == OBJ_ENCODING_ZIPLIST) {
             it->zl.zl = op->subject->ptr;
-            it->zl.eptr = ziplistIndex(it->zl.zl,0);
+            it->zl.eptr = ziplistIndex(it->zl.zl,-2);
             if (it->zl.eptr != NULL) {
                 it->zl.sptr = ziplistNext(it->zl.zl,it->zl.eptr);
```

Comment:
i think we may be iterating on the last element twice (please check).
i.e. we start with the element name of the last element (second to the last in the ziplist).
then we go forward to take it's score.
but then we'll use zzlPrev to to go backwards, which will call ziplistPrev twice (serving the same element we just saw again).
if i'm right, the fix is to replace these two lines with ziplistIndex(-1), and ziplistPrev (switching eptr with sptr)

Reply:
Initially I implemented exactly as you said, and it worked, but then I realized that there is no need, just changing the initial index already does what we want. What matters for zzlPrev is that eptr points to the first element in the "pair", and sptr the second element. That's why it calls ziplistPrev twice, going back from eptr, the first call to ziplistPrev gets the previous sptr, and then gets the previous eptr.

In the initialization, we first get to eptr by getting the element just before the last one. The call to ziplistNext then gets the tail element, which is just after eptr. I can reverse the order like you said (get sptr at "-1", then eptr by using ziplistPrev), if you think it makes the code more readable, but it works the same.

Reply:
@felipou maybe we don't understand each other.
i don't care if we go to -2, then -1, or go to -1 and then -2.
but i think with the current code we iterate on the last item in the set twice.
i.e. in our first iteration, eptr is -2, and sptr is -1.
then we call zzlPrev, which will set sptr to -1, and etpr to -2.
then call zzlPrev again, which will set sptr to -3 and eptr to -4.

or am i missing something?

Reply:
We init eptr as -2 and sptr -1. The zuiNext call will return these values, and also call zzlPrev. The zzlPrev call will call ziplistPrev on eptr (-2) to get the new sptr at -3, and then it calls ziplistPrev again but now on the new sptr (-3), getting the new eptr at -4.

The ziplistPrev gets an element as an argument, and returns the elements before it. So if we pass the eptr at -2, we will get the previous sptr at -3.

Reply:
ok got it.
thanks.